### DISCUSSION TITLE
POST an auth_key parameter for authentication

### DISCUSSION THREAD

**Discussion by locri3n** [grafana/grafana-json-datasource#234] [Q&A] [answered]
Hello,

I have an JSON endpoint that I am trying to hit that requires me to POST an auth_key.

I can curl the information and get a proper response like this:
curl -d "auth_key=12345-1234-4321-54321-abdcefg&limit=100" -X POST https://subdomain.domain.com/other/stuff

My issue is whenever I try to put the auth_key in any of the the available locations in the plugin it comes out NULL on the other end.

Locations I have put the auth_key into:

Datasource -> Custom HTTP Headers
Datasource -> Query string

Panel -> 
  Field = $ (JSON path parent selector)
  Path = POST /endpoint/location (or blank for status request which GETs fine)
  Params = auth_key / valueofkey
  Headers = auth_key / valueofkey

I have implemented these endpoints with fetch requests in JavaScript without any issues so I know my endpoints are solid.

I have tried all these options but I keep getting a NULL value on the other side. (data logged in Python script)
Is there something I am missing that is preventing the auth_key value from posting?

Thanks for supporting the community,
Michael

**Comment by marcusolsson** [comment]
That should indeed work. If none of those places work, it sounds like the value could actually be NULL/empty? Are you getting the value from a variable? Are you able to add any other values?

**Reply by locri3n** [reply]
I was able to resolve the issue. Thanks for getting back to me so quick though.

**Comment by locri3n** [comment, marked as answer]
I was able to resolve this issues once I started posting the data to https://httpbin.org/post to get a better idea of how the data was being formatted.

The main issue for me was that the POST data was being send in [args] rather than [form]:
{
  "args": { "bob": "dog" },
  "data": "",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "application/json, text/plain, */*",
    "Accept-Encoding": "gzip, deflate",
    "Accept-Language": "en-US,en;q=0.5",
    "Content-Length": "0",
    "Content-Type": "application/json",
    "Host": "httpbin.org",
    "User-Agent": "Grafana/8.3.3",
    "X-Amzn-Trace-Id": "Root=12345",
    "X-Grafana-Org-Id": "1"
  },
  "json": null,
  "origin": "ip.addresses",
  "url": "https://httpbin.org/post?bob=dog"
}

How I wanted the data:
{
  "args": {},
  "data": "",
  "files": {},
  "form": {
    "bob": "dog"
  },
  "headers": {
    "Accept-Encoding": "gzip",
    "Content-Length": "7",
    "Content-Type": "application/x-www-form-urlencoded",
    "Host": "httpbin.org",
    "User-Agent": "Grafana",
    "X-Amzn-Trace-Id": "Root=12345"
  },
  "json": null,
  "origin": "ip.address",
  "url": "https://httpbin.org/post"
}

The key to getting the data into form was to set the Content-Type in the data source headers to application/x-www-form-urlencoded.

Once I did that I was able to put the bob=dog value into the "Method and Body" under the "HTTP method, Query param, Headers" option inside the actual query panel.

Hopefully this helps someone else who runs into this issue!

**Comment by nafgarda** [comment]
I have a similar situation. I am not even able to add the API as a data source ussing Grafana JSON API
This is the endpoint https://shelly-30-eu.shelly.cloud. I always get "JSON API: Cannot connect to API"

This the way it works with curl:
curl -X POST https://<server_uri>/device/status -d "id=<device_id>&auth_key=<auth_key>"

Regards
